### ISSUE TITLE
Refactor savegames to support level select and level restart properly

### ISSUE THREAD

**Issue by walkawayy** [rr-/Tomb1Main#499]
Currently, level selecting the same level as the savegame or restarting the level can carry over lost health or items from the current savegame.

A `LEVEL_END_INFO` needs to be added to go along with `LEVEL_START_INFO`, add `GFL_RESTART`, add `GF_RESTART_GAME`, etc.

See discussion in #494.

**Comment by walkawayy** [comment]
>So restarting a level or level select would use level - 1's END_INFO?
from rr:

Yes

Something like this:

```c
struct LEVEL_COMMON_INFO {
    int health;
    ammo etc.;
};

struct LEVEL_START_INFO {
    LEVEL_COMMON_INFO common;
};

struct LEVEL_END_INFO {
    LEVEL_COMMON_INFO common;
    LEVEL_STATS stats;
};
```

Move `Savegame_PersistGameToStartInfo` to `Savegame_PersistGameToEndInfo`.
Delete `Savegame_PersistGameToStartInfo` because it will no longer be needed.

in `StopGame`:

```diff
-        Savegame_PersistGameToStartInfo(g_CurrentLevel + 1);
+        Savegame_CarryEndInfoToStartInfo(g_CurrentLevel, g_CurrentLevel + 1);
         Savegame_ApplyLogicToStartInfo(g_CurrentLevel + 1);
```

In savegame.c create a function that roughly does
```c
void Savegame_CarryEndInfoToStartInfo(int32_t src_level, int32_t dst_level)
{
    g_GameInfo.start[dst_level].common = g_GameInfo.end[src_level].common;
}
```

Refactor savegame_legacy.c to write things based on `g_GameInfo.end[].common` instead of `g_GameInfo.start[]`
Refactor savegame_legacy.c to duplicate the legacy savegame START_INFO data to both structs
Refactor savegame_bson.c to persist both of the new structs
Refactor savegame_bson.c to read both of the new structs, and to maintain backwards compatibility, if no end_info exists in a file, copy it over from start_info

Introduce new GFL_ flag, GFL_RESTART, and new GF_ flag, GF_RESTART_GAME. Have GF_RESTART_GAME run the same code as GF_START_SAVED_GAME except use GFL_RESTART instead of GFL_SAVED.

Go through all usages of GFL_SAVED and amend to `|| GFL_RESTART` if necessary.

Add a flag to `GAME_INFO` that controls whether `InitialiseLaraInventory` should read from `START_INFO` or `END_INFO`.
Refactor `InitialiseLaraInventory` to read from either depending on that flag.
In `StartGame` set that flag `flag = level_type == GFL_SAVED` (but **not** `GFL_RESTART`, this distinction is what we introduce this level type for). I'm OK with having an equivalent of this logic in savegame.c. I'm also OK with storing `current_level_type` instead of a boolean flag which should be more versatile.

- [x] Restarting the level after death should work. With `GFL_RESTART` `InitialiseLaraInventory` will read from `START_INFO`. And in the new approach we never modify START_INFO upon saving, only between levels. 
- [X] Select level feature should work for similar reasons as above.
- [x] Continuing into next level works thanks to `Savegame_PersistEndInfoToStartInfo` and thanks to `GFL_NORMAL`, `InitialiseLaraInventory` will read from `START_INFO`.
- [x] Loading the game should continue to work by setting the flag to `1` due to `GFL_SAVED`, which means `InitialiseLaraInventory` will read from `END_INFO` instead that contains the most recent inventory state.
- [x] Starting new game should work becuase `GAME_INFO` is zero-initialized and `GFL_NORMAL` instructs `InitialiseLaraLoad` to read from `START_INFO`.
- [x] Loading legacy saves should just continue to exhibit the bugs that we already had and continue to work otherwise.
- [x] Loading new saves made prior to this refactor should behave as above.
- [x] Loading new saves made after this refactor should work without any bugs.

If you're up to the task please do these changes in a separate PR; we'll need to have the testers test the heck out of the refactors.

**Comment by rr-** [comment]
@walkawayy I think this can get closed after 2. right?

**Comment by walkawayy** [comment]
Yes.

**Comment by walkawayy** [comment]
Closed in #494.
